File: ios/Classes/OtaUpgradePlugin.swift
import Flutter
import UIKit
import ota_upgrade 
import CoreBluetooth

@objc public class OtaUpgradePlugin: NSObject, FlutterPlugin {
  // Telink OTA related channels
  private var telinkMethodChannel: FlutterMethodChannel!
  private var telinkStatusEventChannel: FlutterEventChannel!
  private var telinkProgressEventChannel: FlutterEventChannel!
  
  // Main plugin channel
  private var mainPluginMethodChannel: FlutterMethodChannel!

  private var statusStreamHandler = StatusStreamHandler()
  private var progressStreamHandler = ProgressStreamHandler()
  private var registrar: FlutterPluginRegistrar!
  
  // 简化：只跟踪当前OTA任务
  private var currentOtaTask: String? = nil
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = OtaUpgradePlugin()
    instance.registrar = registrar
    
    // 初始化 Telink OTA Channels
    instance.telinkMethodChannel = FlutterMethodChannel(
      name: "com.example.ota_upgrade/telink_ota_method", 
      binaryMessenger: registrar.messenger()
    )
    instance.telinkStatusEventChannel = FlutterEventChannel(
      name: "com.example.ota_upgrade/telink_ota_status_event", 
      binaryMessenger: registrar.messenger()
    )
    instance.telinkProgressEventChannel = FlutterEventChannel(
      name: "com.example.ota_upgrade/telink_ota_progress_event", 
      binaryMessenger: registrar.messenger()
    )
    
    registrar.addMethodCallDelegate(instance, channel: instance.telinkMethodChannel)
    instance.telinkStatusEventChannel.setStreamHandler(instance.statusStreamHandler)
    instance.telinkProgressEventChannel.setStreamHandler(instance.progressStreamHandler)

    // 初始化主插件 Channel
    instance.mainPluginMethodChannel = FlutterMethodChannel(
      name: "ota_upgrade", 
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: instance.mainPluginMethodChannel)

    print("OtaUpgradePlugin: All channels registered.")
    
    // 确保状态干净
    instance.currentOtaTask = nil
    
    // 重置底层的OTA状态
    let wrapper = TelinkOtaWrapper.sharedInstance()
    wrapper.cancelOta()
    
    print("OtaUpgradePlugin: 初始化完成，状态已重置")
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    print("OtaUpgradePlugin: 收到方法调用: \(call.method)")
    
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    
    case "startTelinkOta", "startTelinkOtaIOS":
      handleStartOta(call: call, result: result)
    
    case "cancelOta":
      handleCancelOta(result: result)
      
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  private func handleStartOta(call: FlutterMethodCall, result: @escaping FlutterResult) {
    print("OtaUpgradePlugin: 处理 startOta")
    
    // 检查是否已有OTA在进行
    if currentOtaTask != nil {
      result(FlutterError(code: "BUSY", message: "另一个 OTA 进程正在进行中: \(currentOtaTask!)", details: nil))
      return
    }
    
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数格式错误", details: nil))
      return
    }
    
    let macAddress = args["macAddress"] as? String
    let deviceName = args["deviceName"] as? String  
    let filePath = args["filePath"] as? String
    let isAsset = args["fileInAsset"] as? Bool ?? false
    let readInterval = args["readInterval"] as? Int ?? 8
    
    // 参数验证
    if filePath == nil {
      result(FlutterError(code: "INVALID_ARGS", message: "固件文件路径缺失", details: nil))
      return
    }
    
    let deviceId = macAddress ?? deviceName
    if deviceId == nil {
      result(FlutterError(code: "INVALID_ARGS", message: "必须提供macAddress或deviceName参数", details: nil))
      return
    }
    
    // 处理文件路径
    if isAsset, let assetPath = filePath {
      loadAssetFile(assetPath) { [weak self] localPath in
        guard let self = self, let localPath = localPath else {
          result(FlutterError(code: "FILE_ERROR", message: "无法加载固件文件", details: nil))
          self?.statusStreamHandler.send(state: "failed", errorMessage: "无法加载固件文件")
          return
        }
        
        self.performOta(deviceId: deviceId!, filePath: localPath, readInterval: readInterval, result: result)
      }
    } else if let path = filePath {
      performOta(deviceId: deviceId!, filePath: path, readInterval: readInterval, result: result)
    }
  }
  
  private func performOta(deviceId: String, filePath: String, readInterval: Int, result: @escaping FlutterResult) {
    print("OtaUpgradePlugin: 执行OTA - 设备: \(deviceId), 文件: \(filePath)")
    
    currentOtaTask = deviceId
    
    // 发送初始状态
    statusStreamHandler.send(state: "connecting", macAddress: deviceId)
    
    let wrapper = TelinkOtaWrapper.sharedInstance()
    
    // 设置目标设备
    wrapper.setConnectedDevice(deviceId)
    
    // 直接调用wrapper的OTA方法，让它处理所有连接和OTA逻辑
    wrapper.startOta(withFilePath: filePath, readInterval: readInterval, 
      progressCallback: { [weak self] progress in
        guard let self = self else { return }
        
        // 发送进度更新
        self.progressStreamHandler.send(progress: progress)
        
        // 第一次进度更新时发送connected和starting状态
        if progress == 0 {
          self.statusStreamHandler.send(state: "connected", macAddress: deviceId)
          self.statusStreamHandler.send(state: "starting", macAddress: deviceId)
        }
        
        // 有进度就发送progress状态
        if progress > 0 {
          self.statusStreamHandler.send(state: "progress", macAddress: deviceId)
        }
      },
      completionCallback: { [weak self] success, errorMessage in
        guard let self = self else { return }
        
        if success {
          self.statusStreamHandler.send(state: "completed", macAddress: deviceId)
        } else {
          let error = errorMessage ?? "OTA失败"
          
          // 分析错误类型，提供更详细的状态
          if error.contains("蓝牙未开启") {
            self.statusStreamHandler.send(state: "failed", macAddress: deviceId, errorMessage: "蓝牙未开启")
          } else if error.contains("连接失败") || error.contains("连接超时") {
            self.statusStreamHandler.send(state: "failed", macAddress: deviceId, errorMessage: "设备连接失败")
          } else if error.contains("无法找到指定设备") {
            self.statusStreamHandler.send(state: "failed", macAddress: deviceId, errorMessage: "未找到设备")
          } else {
            self.statusStreamHandler.send(state: "failed", macAddress: deviceId, errorMessage: error)
          }
        }
        
        // 清理当前任务
        self.currentOtaTask = nil
      }
    )
    
    // 立即返回，实际结果通过EventChannel回调
    result("OTA进程已启动")
  }
  
  private func handleCancelOta(result: @escaping FlutterResult) {
    print("OtaUpgradePlugin: 取消OTA")
    
    let wrapper = TelinkOtaWrapper.sharedInstance()
    wrapper.cancelOta()
    
    statusStreamHandler.send(state: "aborted", macAddress: currentOtaTask)
    currentOtaTask = nil
    
    result("OTA已取消")
  }
  
  private func loadAssetFile(_ assetPath: String, completion: @escaping (String?) -> Void) {
    print("OtaUpgradePlugin: 加载资源文件: \(assetPath)")
    
    let key = registrar.lookupKey(forAsset: assetPath)
    if let path = Bundle.main.path(forResource: key, ofType: nil) {
      completion(path)
      return
    }
    
    let packageKey = registrar.lookupKey(forAsset: assetPath, fromPackage: "ota_upgrade")
    if let path = Bundle.main.path(forResource: packageKey, ofType: nil) {
      completion(path)
      return
    }
    
    // 检查 App.framework
    if let frameworkPath = Bundle.main.path(forResource: "App", ofType: "framework"),
       let appBundle = Bundle(path: frameworkPath) {
      
      if let path = appBundle.path(forResource: key, ofType: nil) {
        completion(path)
        return
      }
      
      if let path = appBundle.path(forResource: packageKey, ofType: nil) {
        completion(path)
        return
      }
    }
    
    print("OtaUpgradePlugin: 无法找到资源文件")
    completion(nil)
  }
}

// MARK: - Stream Handlers
class StatusStreamHandler: NSObject, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }
  
  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }
  
  func send(state: String, macAddress: String? = nil, errorMessage: String? = nil) {
    guard let eventSink = eventSink else { return }
    
    var event: [String: Any] = ["state": state]
    if let macAddress = macAddress {
      event["macAddress"] = macAddress
    }
    if let errorMessage = errorMessage {
      event["errorMessage"] = errorMessage
    }
    
    DispatchQueue.main.async {
      eventSink(event)
    }
  }
}

class ProgressStreamHandler: NSObject, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }
  
  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }
  
  func send(progress: Float) {
    guard let eventSink = eventSink else { return }
    
    DispatchQueue.main.async {
      eventSink(["progress": progress])
    }
  }
}

File: android/src/main/kotlin/com/zyg/ota_upgrade/OtaUpgradePlugin.kt
package com.zyg.ota_upgrade

import androidx.annotation.NonNull
import android.util.Log
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothGatt
import android.content.Context
import android.os.Handler
import android.os.Looper
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream

import com.telink.ota.ble.Device
import com.telink.ota.fundation.OtaSetting
import com.telink.ota.fundation.StatusCode
import com.telink.ota.util.OtaLogger

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.EventChannel
import java.util.*
import android.bluetooth.BluetoothGattCharacteristic

/** OtaUpgradePlugin */
class OtaUpgradePlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var mainChannel: MethodChannel
  private lateinit var telinkOtaChannel: MethodChannel
  private lateinit var telinkStatusChannel: EventChannel
  private lateinit var telinkProgressChannel: EventChannel
  
  private val TAG = "OtaUpgradePlugin"

  // Telink OTA 状态参数
  private var applicationContext: Context? = null
  private val mainThreadHandler = Handler(Looper.getMainLooper())
  private var mDevice: Device? = null
  private var telinkStatusEventSink: EventChannel.EventSink? = null
  private var telinkProgressEventSink: EventChannel.EventSink? = null
  private var currentOtaMacAddress: String? = null

  // 使用固定的写入延迟
  private val WRITE_DELAY = 20L // 毫秒，根据需要调整

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    Log.d(TAG, "onAttachedToEngine called")
    
    // 保存 Context
    applicationContext = flutterPluginBinding.applicationContext
    
    // 1. 设置主通道
    mainChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "ota_upgrade")
    mainChannel.setMethodCallHandler(this)
    
    // 2. 设置 Telink OTA 通道 - 与 telink_ota.dart 中的名称完全一致
    telinkOtaChannel = MethodChannel(flutterPluginBinding.binaryMessenger, 
                                   "com.example.ota_upgrade/telink_ota_method")
    telinkOtaChannel.setMethodCallHandler(this)
    
    // 3. 设置事件通道
    telinkStatusChannel = EventChannel(flutterPluginBinding.binaryMessenger,
                                     "com.example.ota_upgrade/telink_ota_status_event")
    telinkStatusChannel.setStreamHandler(telinkStatusStreamHandler)
    
    telinkProgressChannel = EventChannel(flutterPluginBinding.binaryMessenger,
                                       "com.example.ota_upgrade/telink_ota_progress_event")
    telinkProgressChannel.setStreamHandler(telinkProgressStreamHandler)
    
    // 启用 TelinkOtaLib 日志
    OtaLogger.ENABLE = true
    
    // 4. 打印确认日志
    Log.d(TAG, "所有通道已设置完成，包括 Telink OTA 通道")
  }

  // Telink OTA 状态流处理器
  private val telinkStatusStreamHandler = object : EventChannel.StreamHandler {
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
      Log.d(TAG, "Telink Status EventChannel onListen")
      telinkStatusEventSink = events
    }

    override fun onCancel(arguments: Any?) {
      Log.d(TAG, "Telink Status EventChannel onCancel")
      telinkStatusEventSink = null
    }
  }

  // Telink OTA 进度流处理器
  private val telinkProgressStreamHandler = object : EventChannel.StreamHandler {
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
      Log.d(TAG, "Telink Progress EventChannel onListen")
      telinkProgressEventSink = events
    }

    override fun onCancel(arguments: Any?) {
      Log.d(TAG, "Telink Progress EventChannel onCancel")
      telinkProgressEventSink = null
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    Log.d(TAG, "收到方法调用: ${call.method}")
    
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      "startTelinkOta" -> {
        Log.d(TAG, "收到 startTelinkOta 调用，参数: ${call.arguments}")
        
        val macAddress = call.argument<String>("macAddress")
        val filePath = call.argument<String>("filePath")
        val fileInAsset = call.argument<Boolean>("fileInAsset") ?: false
        val readInterval = call.argument<Int>("readInterval") ?: 8
        val serviceUUID = call.argument<String>("serviceUUID")
        val characteristicUUID = call.argument<String>("characteristicUUID")
        val resumeFromPacketIndex = call.argument<Int>("resumeFromPacketIndex")
        val packetDelayMs = call.argument<Int>("packetDelayMs")

        if (mDevice != null) {
          result.error("BUSY", "另一个 OTA 进程正在进行中: $currentOtaMacAddress", null)
          return
        }
        if (macAddress == null || !BluetoothAdapter.checkBluetoothAddress(macAddress)) {
          result.error("INVALID_ARGS", "无效的 MAC 地址: $macAddress", null)
          return
        }
        if (filePath == null) {
          result.error("INVALID_ARGS", "固件文件路径不能为空", null)
          return
        }

        try {
          // 读取固件文件
          val firmwareData = readFirmwareFile(filePath, fileInAsset)
          if (firmwareData == null || firmwareData.isEmpty()) {
            result.error("FILE_ERROR", "无法读取固件文件或文件为空", null)
            return
          }
          
          currentOtaMacAddress = macAddress
          startFullOtaProcess(macAddress, firmwareData, readInterval, serviceUUID, characteristicUUID, resumeFromPacketIndex, packetDelayMs)
          result.success("OTA 进程已为 $macAddress 启动")
        } catch (e: Exception) {
          Log.e(TAG, "读取固件文件失败", e)
          result.error("FILE_ERROR", "读取固件文件失败: ${e.message}", null)
        }
      }
      "cancelOta" -> {
        Log.d(TAG, "收到 cancelOta 调用")
        if (mDevice != null) {
          sendStatusUpdate("cancelling", currentOtaMacAddress, null)
          cleanupCurrentDevice(true) // 停止并断开连接
          sendStatusUpdate("aborted", currentOtaMacAddress, null)
          result.success("已请求取消 OTA: $currentOtaMacAddress")
        } else {
          result.error("IDLE", "当前没有活动的 OTA 进程可取消", null)
        }
      }
      else -> {
        Log.w(TAG, "未知方法: ${call.method}")
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    Log.d(TAG, "onDetachedFromEngine called")
    mainChannel.setMethodCallHandler(null)
    telinkOtaChannel.setMethodCallHandler(null)
    telinkStatusChannel.setStreamHandler(null)
    telinkProgressChannel.setStreamHandler(null)
    cleanupCurrentDevice(true)
    applicationContext = null
  }

  // --- Telink OTA 辅助方法 ---
  private fun sendStatusUpdate(state: String, macAddress: String?, errorMessage: String?) {
    telinkStatusEventSink?.let { sink ->
      val event = mutableMapOf<String, Any?>()
      event["state"] = state
      macAddress?.let { event["macAddress"] = it }
      errorMessage?.let { event["errorMessage"] = it }
      mainThreadHandler.post { sink.success(event) }
      Log.d(TAG, "发送状态更新: $state${errorMessage?.let { " 错误: $it" } ?: ""}")
    } ?: Log.w(TAG, "无法发送状态更新，telinkStatusEventSink 为 null")
  }

  private fun sendProgressUpdate(progress: Int, macAddress: String?) {
    telinkProgressEventSink?.let { sink ->
      val event = mutableMapOf<String, Any?>()
      event["progress"] = progress
      // 附带分包级别的断点信息
      mDevice?.let { dev ->
        try {
          event["packetIndex"] = dev.getCurrentPacketIndex()
          event["totalPackets"] = dev.getTotalPackets()
        } catch (_: Exception) {}
      }
      macAddress?.let { event["macAddress"] = it }
      mainThreadHandler.post { sink.success(event) }
    } ?: Log.w(TAG, "无法发送进度更新，telinkProgressEventSink 为 null")
  }

  // OTA 进程暂存数据 (因为连接是异步的)
  private var pendingFirmwareData: ByteArray? = null
  private var pendingReadInterval: Int = 8
  private var pendingServiceUUID: UUID? = null
  private var pendingCharacteristicUUID: UUID? = null
  private var pendingResumeFromPacketIndex: Int? = null
  private var pendingPacketDelayMs: Int? = null

  private fun startFullOtaProcess(
    macAddress: String,
    firmwareData: ByteArray,
    readInterval: Int,
    serviceUUIDStr: String?,
    characteristicUUIDStr: String?,
    resumeFromPacketIndex: Int?,
    packetDelayMs: Int?
  ) {
    Log.d(TAG, "开始 OTA 进程: $macAddress")
    val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
    if (bluetoothAdapter == null) {
      sendStatusUpdate("failed", macAddress, "蓝牙不可用")
      cleanupCurrentDevice(false)
      return
    }

    try {
      val bluetoothDevice = bluetoothAdapter.getRemoteDevice(macAddress)
      applicationContext?.let { ctx ->
        mDevice = Device(ctx)
        mDevice?.setDeviceStateCallback(otaDeviceCallback)
        sendStatusUpdate("connecting", macAddress, null)
        mDevice?.connect(bluetoothDevice)

        // 保存固件数据和参数，以便连接成功后使用
        this.pendingFirmwareData = firmwareData
        this.pendingReadInterval = readInterval
        this.pendingServiceUUID = serviceUUIDStr?.let { UUID.fromString(it) }
        this.pendingCharacteristicUUID = characteristicUUIDStr?.let { UUID.fromString(it) }
        // 把断点信息也暂存下来（通过 OtaSetting 传递）
        this.pendingResumeFromPacketIndex = resumeFromPacketIndex
        // 暂存自定义包间延迟
        this.pendingPacketDelayMs = packetDelayMs

      } ?: run {
        Log.e(TAG, "Context 为 null，无法启动 OTA")
        sendStatusUpdate("failed", macAddress, "内部错误: Context 为 null")
        cleanupCurrentDevice(false)
      }

    } catch (e: IllegalArgumentException) {
      Log.e(TAG, "startFullOtaProcess 中的 IllegalArgumentException: ${e.message}", e)
      sendStatusUpdate("failed", macAddress, "无效的 MAC 地址格式: ${e.message}")
      cleanupCurrentDevice(false)
    } catch (e: Exception) {
      Log.e(TAG, "startFullOtaProcess 中的异常: ${e.message}", e)
      sendStatusUpdate("failed", macAddress, "启动连接失败: ${e.message}")
      cleanupCurrentDevice(false)
    }
  }

  // 设备状态回调
  private val otaDeviceCallback = object : Device.DeviceStateCallback {
    override fun onConnectionStateChange(device: Device, state: Int) {
      val mac = device.macAddress
      Log.d(TAG, "回调 onConnectionStateChange: $state for $mac")

      if (mac == null || mac != currentOtaMacAddress) {
        if (mDevice == null) {
          Log.w(TAG, "忽略过期的 onConnectionStateChange，mDevice 为 null")
          return
        }
        Log.w(TAG, "收到意外设备的连接状态变化: $mac")
        return
      }

      when (state) {
        BluetoothGatt.STATE_CONNECTING -> {
          Log.d(TAG, "设备正在连接: $mac")
        }
        BluetoothGatt.STATE_CONNECTED -> {
          Log.d(TAG, "设备已连接: $mac，正在发现服务...")
          sendStatusUpdate("discoveringServices", mac, null)
          
          // 连接后启动 OTA
          pendingFirmwareData?.let { fwData ->
            val setting = OtaSetting().apply {
              setFirmwareData(fwData)
              setReadInterval(pendingReadInterval)
              pendingServiceUUID?.let { setServiceUUID(it) }
              pendingCharacteristicUUID?.let { setCharacteristicUUID(it) }
            }
            // 直接设置断点续传分片索引
            setting.setResumeFromPacketIndex(pendingResumeFromPacketIndex)
            // 如果设置了自定义包延迟，尝试注入到 Device（通过反射设置字段）
            try {
              pendingPacketDelayMs?.let { delay ->
                val field = Device::class.java.getDeclaredField("PACKET_DELAY_MS")
                field.isAccessible = true
                field.setInt(mDevice, delay)
              }
            } catch (_: Exception) {}
            Log.d(TAG, "尝试使用配置启动 OTA 进程")
            mDevice?.startOta(setting)
          } ?: run {
            Log.e(TAG, "已连接但无固件数据启动 OTA!")
            sendStatusUpdate("failed", mac, "内部错误: 连接后无固件数据")
            cleanupCurrentDevice(true)
          }
        }
        BluetoothGatt.STATE_DISCONNECTING -> {
          Log.d(TAG, "设备正在断开连接: $mac")
        }
        BluetoothGatt.STATE_DISCONNECTED -> {
          Log.d(TAG, "设备已断开连接: $mac")
          var connectionStatus = -1  // 定义在外部作用域的状态变量
          
          if (device != null && device.javaClass.methods.any { it.name == "getLastGattStatus" }) {
            try {
              // 尝试获取详细错误码（如果 Device 类提供）
              val statusMethod = device.javaClass.getMethod("getLastGattStatus")
              connectionStatus = (statusMethod.invoke(device) as? Int) ?: -1
              Log.d(TAG, "断开连接的详细错误码: $connectionStatus")
            } catch (e: Exception) {
              Log.e(TAG, "获取错误码失败: ${e.message}")
            }
          }
          
          if (mDevice != null) {
            if (connectionStatus == 133) {  // 使用外部作用域的变量
              Log.w(TAG, "设备连接错误 (错误码 133): 设备可能不存在或不在范围内")
              sendStatusUpdate("failed", mac, "设备连接错误: 设备不存在或不在范围内")
            } else {
              Log.w(TAG, "连接意外中断: $mac")
              sendStatusUpdate("failed", mac, "连接意外中断")
            }
            cleanupCurrentDevice(false)
          } else {
            Log.d(TAG, "$mac 的断开回调，但 mDevice 为 null (已清理)")
          }
        }
      }
    }

    override fun onOtaStateChanged(device: Device, statusCode: StatusCode) {
      val mac = device.macAddress
      Log.d(TAG, "回调 onOtaStateChanged: ${statusCode.desc} for $mac")

      if (mac == null || mac != currentOtaMacAddress) {
        if (mDevice == null && statusCode != StatusCode.SUCCESS && statusCode != StatusCode.STOPPED && !statusCode.isFailed()) {
          Log.w(TAG, "忽略过期的 onOtaStateChanged，mDevice 为 null")
          return
        }
        Log.w(TAG, "收到意外设备的 OTA 状态更新: $mac / $statusCode")
        return
      }

      when (statusCode) {
        StatusCode.STARTED -> {
          Log.i(TAG, "$mac 的 OTA 已开始")
          sendStatusUpdate("starting", mac, null)
          sendStatusUpdate("progress", mac, null)
        }
        StatusCode.SUCCESS -> {
          Log.i(TAG, "$mac 的 OTA 成功")
          sendStatusUpdate("completed", mac, null)
          cleanupCurrentDevice(true)
        }
        StatusCode.STOPPED -> {
          Log.i(TAG, "$mac 的 OTA 已停止")
          sendStatusUpdate("aborted", mac, null)
          if (mDevice != null) cleanupCurrentDevice(true)
        }
        else -> {
          if (statusCode.isFailed()) {
            Log.e(TAG, "$mac 的 OTA 失败，状态: ${statusCode.desc}")
            sendStatusUpdate("failed", mac, statusCode.desc)
            cleanupCurrentDevice(true)
          } else {
            Log.w(TAG, "$mac 的未处理 OTA 状态: ${statusCode.desc}")
          }
        }
      }
    }

    override fun onOtaProgressUpdate(progress: Int) {
      sendProgressUpdate(progress, currentOtaMacAddress)
    }
  }

  private fun cleanupCurrentDevice(disconnect: Boolean) {
    Log.d(TAG, "清理设备: $currentOtaMacAddress, 断开连接: $disconnect")
    mDevice?.let { dev ->
      Log.d(TAG, "清理 $currentOtaMacAddress 的设备实例")
      dev.setDeviceStateCallback(null)
      dev.clearAll(disconnect)
    }
    mDevice = null
    currentOtaMacAddress = null
    pendingFirmwareData = null
    pendingServiceUUID = null
    pendingCharacteristicUUID = null
    Log.d(TAG, "设备清理完成")
  }

  // 添加读取固件文件的方法
  private fun readFirmwareFile(filePath: String, isAsset: Boolean): ByteArray? {
    Log.d(TAG, "读取固件文件: $filePath, 是否为资产: $isAsset")
    var inputStream: InputStream? = null
    
    try {
      if (isAsset) {
        // 从 Flutter 资产中读取
        // 注意: Flutter资产路径通常以 "flutter_assets/" 开头
        val assetManager = applicationContext?.assets
        if (assetManager == null) {
          Log.e(TAG, "无法访问 AssetManager")
          return null
        }
        
        // 处理可能的包前缀
        var assetPath = filePath
        if (assetPath.startsWith("packages/")) {
          // 保留 packages/ 后面的路径部分
          assetPath = "flutter_assets/$assetPath"
        } else {
          assetPath = "flutter_assets/$assetPath"
        }
        
        Log.d(TAG, "尝试从资产加载: $assetPath")
        inputStream = assetManager.open(assetPath)
      } else {
        // 从文件系统读取
        val file = File(filePath)
        if (!file.exists() || !file.isFile) {
          Log.e(TAG, "文件不存在: $filePath")
          return null
        }
        inputStream = FileInputStream(file)
      }
      
      return inputStream?.readBytes()
    } catch (e: IOException) {
      Log.e(TAG, "读取文件失败: ${e.message}", e)
      return null
    } finally {
      inputStream?.close()
    }
  }

  // 在数据包发送后添加延迟
  private fun sendOtaPacket(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, data: ByteArray) {
    // 写入数据
    characteristic.setValue(data)
    gatt.writeCharacteristic(characteristic)
    
    // 固定延迟
    Thread.sleep(WRITE_DELAY)
  }
}
